logical/join.go

```diff
@@ -0,0 +1,226 @@
+		outNodes = make([]physical.Node, len(sourceShuffled))
+
+		for i := range outNodes {
+			outNodes[i] = physical.NewLookupJoin(sourceShuffled[i], joinedShuffled[i], isLeftJoin)
+		}
+	}
+
+	return outNodes, variables, nil
+}
+
+func isConjunctionOfEqualities(f physical.Formula) bool {
+	switch f := f.(type) {
+	case *physical.And:
+		return isConjunctionOfEqualities(f.Left) && isConjunctionOfEqualities(f.Right)
+	case *physical.Predicate:
+		return f.Relation == physical.Equal
+	case *physical.Constant:
+		return f.Value
+	default:
+		return false
+	}
+}
+
+func getKeysFromFormula(formula physical.Formula, sourceNamespace, joinedNamespace *metadata.Namespace) ([]physical.Expression, []physical.Expression, error) {
+	sourceKey := make([]physical.Expression, 0)
+	joinedKey := make([]physical.Expression, 0)
+
+	for _, element := range formula.SplitByAnd() {
+		if _, ok := element.(*physical.Constant); ok {
+			continue
+		}
+
+		predicate := element.(*physical.Predicate) // this won't error since we only have predicates and constants
+
+		leftExpression := predicate.Left
+		rightExpression := predicate.Right
+
+		doesLeftMatchSource := leftExpression.DoesMatchNamespace(sourceNamespace)
+		doesLeftMatchJoined := leftExpression.DoesMatchNamespace(joinedNamespace)
+		doesRightMatchSource := rightExpression.DoesMatchNamespace(sourceNamespace)
+		doesRightMatchJoined := rightExpression.DoesMatchNamespace(joinedNamespace)
+
+		// We check for errors. For example a predicate a.x - b.y = 0 won't have any match on the left formula
+		if !doesLeftMatchSource && !doesLeftMatchJoined {
+			return nil, nil, errors.New("Left expression of predicate doesn't match either sources namespace")
+		} else if !doesRightMatchSource && !doesRightMatchJoined {
+			return nil, nil, errors.New("Right expression of predicate doesn't match either sources namespace")
+		} else if !doesLeftMatchSource && !doesRightMatchSource {
+			return nil, nil, errors.New("Neither side of predicate matches sources namespace")
+		} else if !doesLeftMatchJoined && !doesRightMatchJoined {
+			return nil, nil, errors.New("Neither side of predicate matches joined sources namespace")
+		}
+
+		// Now we want to decide which part of the predicate belong to which key
+		if !doesLeftMatchSource {
+			// If left doesn't match source, then right surely does and also left must match joined
+			sourceKey = append(sourceKey, rightExpression)
+			joinedKey = append(joinedKey, leftExpression)
+		} else if !doesLeftMatchJoined {
+			// Likewise if left doesn't match joined, then right surely does and also left must match source
+			sourceKey = append(sourceKey, leftExpression)
+			joinedKey = append(joinedKey, rightExpression)
+		} else if !doesRightMatchSource {
+			// Left matches both, but right doesn't match source, so right is joined, left is source
+			sourceKey = append(sourceKey, leftExpression)
+			joinedKey = append(joinedKey, rightExpression)
+		} else {
```

Comment:
That's a bit complicated - basically there are two approaches: one can either use only the namespace of the `source` and `joined` nodes, or can also look at variables from respecive `Physical()` calls. If it's the first one, then no, only one match is possible, because all variables are prefixed, so you would have to join two nodes with the same prefix (don't do that...). If you include variables from the `Physical()` calls then indeed, both expressions can match both namespaces, but that means they are both constants and either someone wrote `...JOIN ... ON ... 3 = 3`, in which case nothing interesting happens, or `... JOIN ... ON ... 3 = 4` and then, well..., nothing happens at all. So short answer: it is possible, but then it doesn't matter to which key we assign which expression :D

Reply:
Thanks for an explanation